File: repro/doltlite-fts5-corruption.py
#!/usr/bin/env python3

from __future__ import annotations

import os
import sqlite3
from pathlib import Path

ROOT = Path("/data/projects/claude-context-mode")
MAX_CHUNK_BYTES = 4096
INPUTS = [
    "src/db-base.ts",
    "src/runtime.ts",
    "src/session/extract.ts",
    "src/session/snapshot.ts",
]


def build_title(heading_stack: list[dict[str, object]], current_heading: str) -> str:
    titles = [str(h["text"]) for h in heading_stack]
    return " > ".join(titles) if titles else (current_heading or "untitled")


def chunk_markdown(text: str, max_chunk_bytes: int = MAX_CHUNK_BYTES) -> list[dict[str, object]]:
    chunks: list[dict[str, object]] = []
    lines = text.split("\n")
    heading_stack: list[dict[str, object]] = []
    current_content: list[str] = []
    current_heading = ""

    def flush() -> None:
        nonlocal current_content
        joined = "\n".join(current_content).strip()
        if not joined:
            return

        title = build_title(heading_stack, current_heading)
        has_code = any(line.startswith("```") for line in current_content)

        if len(joined.encode("utf-8")) <= max_chunk_bytes:
            chunks.append({"title": title, "content": joined, "has_code": has_code})
            current_content = []
            return

        paragraphs = joined.split("\n\n")
        accumulator: list[str] = []
        part_index = 1

        def flush_accumulator() -> None:
            nonlocal accumulator, part_index
            if not accumulator:
                return
            part = "\n\n".join(accumulator).strip()
            if not part:
                return
            part_title = f"{title} ({part_index})" if len(paragraphs) > 1 else title
            part_index += 1
            chunks.append(
                {
                    "title": part_title,
                    "content": part,
                    "has_code": "```" in part,
                }
            )
            accumulator = []

        for para in paragraphs:
            accumulator.append(para)
            candidate = "\n\n".join(accumulator)
            if len(candidate.encode("utf-8")) > max_chunk_bytes and len(accumulator) > 1:
                accumulator.pop()
                flush_accumulator()
                accumulator = [para]

        flush_accumulator()
        current_content = []

    i = 0
    while i < len(lines):
        line = lines[i]

        if line and set(line) <= {"-", "_", "*"} and len(line) >= 3:
            flush()
            i += 1
            continue

        if line.startswith("#"):
            level = 0
            while level < len(line) and line[level] == "#":
                level += 1
            if 1 <= level <= 4 and len(line) > level and line[level] == " ":
                flush()
                heading = line[level + 1 :].strip()
                while heading_stack and int(heading_stack[-1]["level"]) >= level:
                    heading_stack.pop()
                heading_stack.append({"level": level, "text": heading})
                current_heading = heading
                current_content.append(line)
                i += 1
                continue

        if line.startswith("```"):
            fence = line.strip()
            code_lines = [line]
            i += 1
            while i < len(lines):
                code_lines.append(lines[i])
                if lines[i].strip() == fence:
                    i += 1
                    break
                i += 1
            current_content.extend(code_lines)
            continue

        current_content.append(line)
        i += 1

    flush()
    return chunks


def main() -> int:
    db_path = "/tmp/doltlite-python-repro.db"
    for suffix in ("", "-wal", "-shm"):
        try:
            os.unlink(db_path + suffix)
        except FileNotFoundError:
            pass

    con = sqlite3.connect(db_path)
    try:
        con.executescript(
            """
            CREATE VIRTUAL TABLE chunks USING fts5(
              title,
              content,
              source_id UNINDEXED,
              content_type UNINDEXED,
              tokenize='porter unicode61'
            );
            """
        )

        for file_index, rel in enumerate(INPUTS, start=1):
            text = (ROOT / rel).read_text(encoding="utf-8")
            chunks = chunk_markdown(text)
            print(f"FILE {rel} chunks {len(chunks)}")
            for chunk_index, chunk in enumerate(chunks, start=1):
                rowid = file_index * 1_000_000 + chunk_index
                content_type = "code" if chunk["has_code"] else "prose"
                try:
                    con.execute(
                        """
                        INSERT INTO chunks (rowid, title, content, source_id, content_type)
                        VALUES (?, ?, ?, ?, ?)
                        """,
                        (
                            rowid,
                            chunk["title"],
                            chunk["content"],
                            file_index,
                            content_type,
                        ),
                    )
                    print(f"  ok {chunk_index} {len(str(chunk['content']).encode('utf-8'))}")
                except Exception as exc:
                    print(f"  fail {chunk_index} {exc}")
                    return 1

        total = con.execute("SELECT count(*) FROM chunks").fetchone()[0]
        print(f"PASS rows={total}")
        return 0
    finally:
        con.close()


if __name__ == "__main__":
    raise SystemExit(main())

File: repro/doltlite-fts5-corruption.mjs
#!/usr/bin/env node

import { mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { spawnSync } from "node:child_process";

const ROOT = "/data/projects/claude-context-mode";
const DOLTLITE = "/data/projects/doltlite-latest/build/doltlite";
const SQLITE3 = "/usr/bin/sqlite3";
const MAX_CHUNK_BYTES = 4096;

const INPUTS = [
  "src/db-base.ts",
  "src/runtime.ts",
  "src/session/extract.ts",
  "src/session/snapshot.ts",
];

function sqlString(text) {
  return `'${text.replace(/'/g, "''")}'`;
}

function buildTitle(headingStack, currentHeading) {
  const titles = headingStack.map((h) => h.text);
  return titles.length ? titles.join(" > ") : (currentHeading || "untitled");
}

function chunkMarkdown(text, maxChunkBytes = MAX_CHUNK_BYTES) {
  const chunks = [];
  const lines = text.split("\n");
  const headingStack = [];
  let currentContent = [];
  let currentHeading = "";

  const flush = () => {
    const joined = currentContent.join("\n").trim();
    if (!joined) return;

    const title = buildTitle(headingStack, currentHeading);
    const hasCode = currentContent.some((l) => /^`{3,}/.test(l));

    if (Buffer.byteLength(joined) <= maxChunkBytes) {
      chunks.push({ title, content: joined, hasCode });
      currentContent = [];
      return;
    }

    const paragraphs = joined.split(/\n\n+/);
    let accumulator = [];
    let partIndex = 1;

    const flushAccumulator = () => {
      if (accumulator.length === 0) return;
      const part = accumulator.join("\n\n").trim();
      if (!part) return;
      const partTitle = paragraphs.length > 1 ? `${title} (${partIndex})` : title;
      partIndex++;
      chunks.push({
        title: partTitle,
        content: part,
        hasCode: part.includes("```"),
      });
      accumulator = [];
    };

    for (const para of paragraphs) {
      accumulator.push(para);
      const candidate = accumulator.join("\n\n");
      if (Buffer.byteLength(candidate) > maxChunkBytes && accumulator.length > 1) {
        accumulator.pop();
        flushAccumulator();
        accumulator = [para];
      }
    }
    flushAccumulator();
    currentContent = [];
  };

  let i = 0;
  while (i < lines.length) {
    const line = lines[i];

    if (/^[-_*]{3,}\s*$/.test(line)) {
      flush();
      i++;
      continue;
    }

    const headingMatch = line.match(/^(#{1,4})\s+(.+)$/);
    if (headingMatch) {
      flush();
      const level = headingMatch[1].length;
      const heading = headingMatch[2].trim();
      while (headingStack.length > 0 && headingStack[headingStack.length - 1].level >= level) {
        headingStack.pop();
      }
      headingStack.push({ level, text: heading });
      currentHeading = heading;
      currentContent.push(line);
      i++;
      continue;
    }

    const codeMatch = line.match(/^(`{3,})(.*)?$/);
    if (codeMatch) {
      const fence = codeMatch[1];
      const codeLines = [line];
      i++;
      while (i < lines.length) {
        codeLines.push(lines[i]);
        if (lines[i].startsWith(fence) && lines[i].trim() === fence) {
          i++;
          break;
        }
        i++;
      }
      currentContent.push(...codeLines);
      continue;
    }

    currentContent.push(line);
    i++;
  }

  flush();
  return chunks;
}

function buildFixture() {
  const lines = [
    ".bail on",
    "CREATE VIRTUAL TABLE chunks USING fts5(",
    "  title,",
    "  content,",
    "  source_id UNINDEXED,",
    "  content_type UNINDEXED,",
    "  tokenize='porter unicode61'",
    ");",
  ];

  for (const [fileIndex, rel] of INPUTS.entries()) {
    const text = readFileSync(resolve(ROOT, rel), "utf8");
    const chunks = chunkMarkdown(text);
    lines.push(`-- ${rel}: ${chunks.length} chunks`);
    for (const [chunkIndex, chunk] of chunks.entries()) {
      const rowid = (fileIndex + 1) * 1_000_000 + chunkIndex + 1;
      const sourceId = fileIndex + 1;
      const contentType = chunk.hasCode ? "code" : "prose";
      lines.push(
        `INSERT INTO chunks (rowid, title, content, source_id, content_type) VALUES (` +
        `${rowid}, ${sqlString(chunk.title)}, ${sqlString(chunk.content)}, ${sourceId}, ${sqlString(contentType)});`,
      );
    }
  }

  lines.push("SELECT count(*) FROM chunks;");
  return lines.join("\n") + "\n";
}

function runEngine(binary, dbPath, fixturePath) {
  return spawnSync(binary, [dbPath], {
    input: readFileSync(fixturePath, "utf8"),
    encoding: "utf8",
  });
}

function main() {
  const tmp = mkdtempSync(join(tmpdir(), "doltlite-fts5-repro-"));
  const fixturePath = join(tmp, "fixture.sql");
  const stockDb = join(tmp, "stock.db");
  const doltDb = join(tmp, "doltlite.db");

  try {
    const fixture = buildFixture();
    writeFileSync(fixturePath, fixture);

    const stock = runEngine(SQLITE3, stockDb, fixturePath);
    const dolt = runEngine(DOLTLITE, doltDb, fixturePath);

    const summary = {
      fixturePath,
      inputs: INPUTS,
      stock: {
        binary: SQLITE3,
        status: stock.status,
        stdout: stock.stdout.trim(),
        stderr: stock.stderr.trim(),
      },
      doltlite: {
        binary: DOLTLITE,
        status: dolt.status,
        stdout: dolt.stdout.trim(),
        stderr: dolt.stderr.trim(),
      },
    };

    console.log(JSON.stringify(summary, null, 2));

    if (stock.status !== 0) process.exit(stock.status ?? 1);
    if (dolt.status === 0) {
      console.error("Expected DoltLite repro to fail, but it passed.");
      process.exit(1);
    }
  } finally {
    if (!process.argv.includes("--keep")) {
      rmSync(tmp, { recursive: true, force: true });
    }
  }
}

main();
